unix/w0vncserver/wayland/PixelBuffer.cxx

```diff
@@ -0,0 +1,159 @@
+#include <core/LogWriter.h>
+#include <rfb/VNCServerST.h>
+
+#include "../w0vncserver.h"
+#include "objects/Output.h"
+#include "objects/Display.h"
+#include "objects/ScreencopyManager.h"
+#include "PixelBuffer.h"
+
+using namespace wayland;
+
+static core::LogWriter vlog("WaylandPixelBuffer");
+
+PixelBuffer::PixelBuffer(Display* display, Output* output_,
+                         rfb::VNCServer* server_,
+                         std::function<void()> desktopReadyCallback_)
+  : ScreencopyManager(display, output_), firstFrame(true),
+    shadowFramebuffer(nullptr),
+    desktopReadyCallback(desktopReadyCallback_), server(server_)
+{
+  captureFrame();
+}
+
+PixelBuffer::~PixelBuffer()
+{
+  delete [] shadowFramebuffer;
+}
+
+void PixelBuffer::captureFrame()
+{
+  if (output->getWidth() != (uint32_t)width() ||
+      output->getHeight() != (uint32_t)height()) {
+    if (!firstFrame) {
+      vlog.debug("Detected resize, calling resize()");
+      resize();
+    }
+  }
+
+  ScreencopyManager::captureFrame();
+}
+
+void PixelBuffer::captureFrameDone()
+{
+  // We need to capture our first frame before we know which format
+  // the display is using.
+  // FIXME: Can we query the compositor instead of doing this?
+  if (firstFrame) {
+    try {
+      format = getPixelFormat();
+    } catch (std::runtime_error& e) {
+      fatal_error("Failed to get pixel format: %s", e.what());
+      return;
+    }
+
+    // Utilize a shadow framebuffer, as we have no way of accessing
+    // the screen contents directly.
+    shadowFramebuffer = new uint8_t[output->getWidth() *
+                                    output->getHeight() * (format.bpp / 8)];
+
+    setSize(output->getWidth(), output->getHeight());
+
+    desktopReadyCallback();
+  }
+
+  firstFrame = false;
+
+  syncBuffers();
+
+  ScreencopyManager::captureFrameDone();
+
+  server->add_changed(getDamage());
+
+  captureFrame();
+}
+
+void PixelBuffer::resize()
+{
+  ScreencopyManager::resize();
+
+  delete [] shadowFramebuffer;
+
+  shadowFramebuffer = new uint8_t[output->getWidth() *
+                                  output->getHeight() * (format.bpp / 8)];
+
+  setSize(output->getWidth(), output->getHeight());
+
+  syncBuffers();
+
+  server->setPixelBuffer(this);
+}
+
+void PixelBuffer::syncBuffers()
+{
+  int srcStride;
+  int dstStride;
+  uint8_t* srcBuffer;
+  uint8_t* dstBuffer;
+  pixman_bool_t ret;
+  core::Region region;
+  std::vector<core::Rect> rects;
+
+  region = getDamage();
+
+  if (region.is_empty())
+    region = getRect();
+
+  srcBuffer = getBufferData();
+  srcStride = width();
+  dstBuffer = getBufferRW(getRect(), &dstStride);
+
+  region.get_rects(&rects);
+  for (core::Rect &rect : rects) {
+    ret = pixman_blt((uint32_t*)srcBuffer, (uint32_t*)dstBuffer,
+                     srcStride, dstStride, format.bpp, format.bpp,
+                     rect.tl.x, rect.tl.y, rect.tl.x, rect.tl.y,
+                     rect.width(), rect.height());
+    commitBufferRW(rect);
```

Comment:
There should be a symmetry between `getBufferRW()` and `commitBufferRW()`. So should probably move something around here.